alf/utils/tensorrt_utils.py

```diff
@@ -0,0 +1,390 @@
+        to ONNX using ``torch.onnx.export``.
+
+        There are mainly several preparation steps:
+
+        1. Put ``module`` in eval mode and set requires_grad to False.
+        2. Remove any 'optimizer' attributes in the ``module``, otherwise ``torch.jit.trace``
+           will report errors.
+        3. ONNX assumes flat arg list. So we need to pack the flat args and kwargs
+           back to the original nest structures before calling ``module.method``.
+        4. [Hacks]: make sure all inputs participate in the ONNX graph; make sure
+           outputs contain no duplicate tensors. The reason is that JIT.trace will
+           only trace inputs that are used in the graph, and create inputs for ONNX
+           model. If we have a nested input, the nest leaves will be selected by
+           JIT.trace depending on their usage. For example, if only TimeStep.observation
+           and TimeStep.reward are used, then in the ONNX model, there are only
+           two input nodes. When running the ONNX model, we need to manually
+           extract the corresponding inputs and send only the two to the model.
+           However, it will be difficult for us to know which leaves are actually
+           used in advance. So the hack here is to just assume all inputs are used
+           so we don't have to resolve the correspondence issue.
+        5. Flatten the method output before returning so that ONNX won't flatten it
+           according to its own defined order. Later we will recover the original
+           output structure using ``self.recover_module_output()``.
+
+        Args:
+            module: The module to be exported.
+            method: A method in ``module`` to be exported.
+            example_args: The example args to be used for ``method``. Should be
+                a tuple of args.
+            example_kwargs: The example kwargs to be used for ``method``. Should be
+                a dict of kargs.
+        """
+        super().__init__()
+        _OnnxWrapper._strip_optimizers(module)
+        self._module = hobot_common.freeze_model(module)
+        self._method = functools.partial(method, self._module)
+        self._example_args = example_args
+        self._example_kwargs = example_kwargs
+        # Sometimes we will return Distributions in the output. In this case,
+        # we need to convert them to
+        example_output = self._method(*example_args, **example_kwargs)
+        self._output_params_spec = dist_utils.extract_spec(
+            dist_utils.distributions_to_params(example_output))
+        self._output_spec = dist_utils.extract_spec(example_output)
+
+    @staticmethod
+    def _strip_optimizers(module):
+        """Recursively set all sub-algorithms' optimizers to None as they are not
+        supported by onnx (torch.jit.trace).
+
+        This operation is in-place.
+        """
+        for m in module.modules():
```

Comment:
how is replay buffer handled?

Reply:
> how is replay buffer handled?

For module buffers, they will be treated as part of the graph if used. Just like the model weights. The replay buffer won't be traced usually because they are only for training.